README.md
# minecraftHawkEye
Module for the Minecraft bot, hit 100% of the time with the bow, using the equations of parabolic shooting with friction

This module is based on bot of <a href="https://github.com/PrismarineJS/mineflayer" target="_blank">mineflayer</a> repository

This calculate the trajectory of arrow / snowballs... to ensure which is the best angle to hit

Install:
- Install nodejs >= 14 from nodejs.org
- npm i minecrafthawkeye

Usage: 

First load mineflayer and minecrafthawkeye, and load plugin into mineflayer
```js
const mineflayer = require('mineflayer')
const minecraftHawkEye = require('minecrafthawkeye');
bot.loadPlugin(minecraftHawkEye)
const weapon = 'bow'
```

Now you can request functions
```js
// Get an a entity
// Example get te player entity, If emtpy return first player found
const playerEntity = bot.hawkEye.getPlayer(playername) 

// Auto attack every 1,2 secs to target
bot.hawkEye.autoAttack(target, weapon)

// Stop auto attack
bot.hawkEye.stop()

// Fire only 1 shot
bot.hawkEye.oneshot(target, weapon)

// Otherwise you can get Yaw and Pitch
bot.hawkEye.getMasterGrade(target, speedInVec3, weapon)
// With Yaw and Pitch you can use yourself bot.look(yaw,pitch) adn activate / deactivate main hand to fire


// Similar has oneshot but need yo add manually the Pitch and Way in radians
bot.hawkEye.simplyShot(yaw, pitch)
```

A example in one file:

```js
// ** Grant OP to Archer **
// file: example.js
const mineflayer = require('mineflayer')
const minecraftHawkEye = require('minecrafthawkeye');
const Vec3 = require('Vec3')

const bot = mineflayer.createBot({
    host: process.argv[2] ? process.argv[2] : 'localhost',
    port: process.argv[3] ? parseInt(process.argv[3]) : 25565,
    username: process.argv[4] ? process.argv[4] : 'Archer',
    password: process.argv[5]
})

bot.loadPlugin(minecraftHawkEye)

bot.on('spawn', function() {
    bot.chat('/give ' + bot.username + ' bow{Enchantments:[{id:unbreaking,lvl:3}]} 1')
    bot.chat('/give ' + bot.username + ' minecraft:arrow 300')
    bot.chat('/time set day')
    bot.chat('/kill @e[type=minecraft:arrow]')
    bot.chat('Ready!')

    // Get target for block position, use whatever you need
    const target = bot.hawkEye.getPlayer()
    console.log(target)
    if (!target) {
        return false
    }

    weapon = 'bow'
    // Auto attack every 1,2 secs until target is dead or is to far away
    bot.hawkEye.autoAttack(target, weapon)
        // If you force stop attack use:
        // hawkEye.stop();

    // Use one shot time with calc:
    // bot.hawkEye.oneShot(target);

    // If you want to shot in XYZ position:
    /*
          const blockPosition = {
                  position: new Vec3(244.5, 75.5, -220),
                  isValid: true
              }
          // bot.hawkEye.oneShot(blockPosition);
          // bot.hawkEye.autoAttack(blockPosition);
      */
})
```
All is done, when is attack mode they get best possibility for impact, and shot arrow every 1,2 secs (max power)

## Valid list of weapons or items that can calculate the perfect shot or launch it directly
```
bow
crossbow
snowball
ender_pearl
egg
splash_potion
trident
```

Events:

When but uses autoAttack and the mob is gone (killed o dissapeared)
```js
bot.on('auto_shop_stopped', (target) => {
    ///
})
```

I'm glad I can help you, do we help each other?

# TODOs
No TODOs!
Anyone have more ideas? :D


docs/history.md
## 1.3.0
* Migrate all code to typescript
* Added test server
* Added unitary test with test server

## 1.2.7
* Added tridents! now you can trow tridents with perfect shot
## 1.2.6
* Great improve CPU optimization
    * Add prismarine-world for use a iterator function, check a raycast between each arrow position
    * Added in bow.js example with drawing raycast of each tick arrow position and blockcolision with iterator
## 1.2.5
* Fix grammar
## 1.2.4
* Added a arrowTrajectoryPoints in the result of getMasterGrade function
* Added example of arrow trayecotry preview on bow.js example
## 1.2.3
* Fixed bug with crossbow enchant
## 1.2.2
* Fixed oneShot (break from 1.2.1)
* Adding more supports!
    * Add snowball
    * Add ender pearl
    * Add splash potions
    * Add eggs
    * Added examples example
## 1.2.1
* Optimized crossbow speed, with enchant3 now is a machine gun!
* Optimized autoattack, now have 2 events separatly for optimize getGrades and Calculate if bow/crossbow is ready to shot
* Added crossbow example with enchant 3
    * Caution max echant lvl is 5 because more than 5 the time is negative! Formula => 1250 - ((lvl ? lvl : 0) * 250)
## 1.2
* Update packages dependency
## 1.1.0
* Important: Now getMasterGrade need a weapon for calculate! from: bot.hawkEye.getMasterGrade(bot, target, speed) to  bot.hawkEye.getMasterGrade(bot, target, speed, weapon)
* Adding crossbow support!
* Added example for use with crossbow
* Added in in autoAttack & oneShot default weapon bow (this must be use of example)
## 1.0.8
* Fixed bug on "small" targets (spiders), if the target is mob or player calculates the height and shot in the middle
* Fixed small bug from start arrow position, this cause some times wrong targets
* Added dynamic precission factor, when arrow is very near of target, split calculation to * 5, for avoid wrong grade calc
## 1.0.7
* Added docs/history.md
* No fixed bugs and methods
* Removed rubish code
* Standarized code (npx standard --fix)
## 1.0.6
* From: bot.hawkEye.getMasterGrade(bot, target, speed);
* To: bot.hawkEye.getMasterGrade(target, speed);
## 1.0.5
* Change bot to plugin method
## 1.0.4
* First functional version

src/hawkEye.ts
import { isEntity, OptionsMasterGrade, Projectil, Weapons, weaponsProps } from './types'
import { Vec3 } from 'vec3'
import getMasterGrade, { calculateArrowTrayectory } from './hawkEyeEquations'
import { Entity } from 'prismarine-entity'
import { bot } from './loadBot'

let target: Entity | OptionsMasterGrade
let preparingShot: boolean
let preparingShotTime: number
let prevPlayerPositions: Array<Vec3> = []
let oneShot: boolean
let chargingArrow: boolean
let weapon: Weapons = Weapons.bow
let infoShot: ReturnType<typeof getMasterGrade>

export const autoAttack = (targetToAttack: Entity | OptionsMasterGrade, inputWeapon = Weapons.bow, isOneShot = false) => {
  if (!targetToAttack) {
    return false
  }
  oneShot = isOneShot

  target = targetToAttack
  preparingShot = false
  prevPlayerPositions = []
  weapon = inputWeapon

  bot.on('physicTick', getGrades)
  bot.on('physicTick', autoCalc)
  return true
}

export const stop = () => {

  bot.deactivateItem()
  bot.removeListener('physicTick', getGrades)
  bot.removeListener('physicTick', autoCalc)
  bot.emit('auto_shop_stopped', target)
}

const getGrades = () => {
  if (target === undefined || (isEntity(target) && !target.isValid)) {
    stop()
    return
  }

  if (prevPlayerPositions.length > 10) { prevPlayerPositions.shift() }

  const position = target.position.clone()

  prevPlayerPositions.push(position)

  const speed = new Vec3(0, 0, 0)

  for (let i = 1; i < prevPlayerPositions.length; i++) {
    const pos = prevPlayerPositions[i]
    const prevPos = prevPlayerPositions[i - 1]
    speed.x += pos.x - prevPos.x
    speed.y += pos.y - prevPos.y
    speed.z += pos.z - prevPos.z
  }

  speed.x = speed.x / prevPlayerPositions.length
  speed.y = speed.y / prevPlayerPositions.length
  speed.z = speed.z / prevPlayerPositions.length

  infoShot = getMasterGrade(target, speed, weapon)
}

const autoCalc = async () => {
  const waitTime = weaponsProps[weapon].waitTime

  const slotID = bot.getEquipmentDestSlot('hand')
  if (bot.inventory.slots[slotID] === null || bot.inventory.slots[slotID].name !== weapon) {
    const weaponFound = bot.inventory.items().find(item => item.name === weapon)
    if (weaponFound) {
      try {
        await bot.equip(weaponFound, 'hand')
        return
      } catch (err) {
        await sleep(500)
        return
      }
    } else {
      stop()
      return
    }
  }

  if (!preparingShot) {
    if (['bow', 'crossbow', 'trident'].includes(weapon)) {
      bot.activateItem()
    }
    preparingShot = true
    preparingShotTime = Date.now()
  }

  if (infoShot) {
    bot.look(infoShot.yaw, infoShot.pitch, true)

    if (preparingShot) {
      if (['bow', 'trident'].includes(weapon) && Date.now() - preparingShotTime > waitTime) {
        bot.deactivateItem()
        preparingShot = false
        if (oneShot) {
          stop()
        }
      }

      if (['snowball', 'ender_pearl', 'egg', 'splash_potion'].includes(weapon) && Date.now() - preparingShotTime > waitTime) {
        bot.swingArm('left')
        bot.activateItem()
        bot.deactivateItem()
        preparingShot = false
        if (oneShot) {
          stop()
        }
      }

      if (weapon === 'crossbow') {
        shotCrossbow()
      }
    }
  }
}

const shotCrossbow = () => {
  if (chargingArrow) {
    bot.activateItem()
    bot.deactivateItem()
    chargingArrow = false
    preparingShot = false
    if (oneShot) {
      stop()
    }
    return
  }

  if (bot.heldItem === null) {
    stop()
    return
  }

  // @ts-ignore pending to fix types from core
  const isEnchanted = bot.heldItem.nbt.value.Enchantments ? bot.heldItem.nbt.value.Enchantments.value.value.find(enchant => enchant.id.value === 'quick_charge') : undefined
  const shotIn = 1250 - ((isEnchanted ? isEnchanted.lvl.value : 0) * 250)

  if (weapon === 'crossbow' && !chargingArrow && Date.now() - preparingShotTime > shotIn) {
    bot.deactivateItem()
    chargingArrow = true
  }
}

const sleep = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms))
}


export const detectAim = () => {
  const entities = Object.values(bot.entities)
    // @ts-ignore PR: https://github.com/PrismarineJS/prismarine-entity/pull/55
    .filter((e) => e.type === "player" || (e.type === 'hostile' && e.name === 'skeleton'))

  const calculatedEntityTarget: Record<string, {
    uuid: string,
    name: string,
    prevTrajectory: Array<Vec3>
  }> = {}

  entities
    .forEach((e) => {
      if (!e.uuid) return
      const calc = calculateArrowTrayectory(e.position, 3, e.pitch, e.yaw, Weapons.bow)
      calculatedEntityTarget[e.uuid] = {
        uuid: e.uuid,
        name: e.type === "player" ? e.username ?? '' : e.name ?? '',
        prevTrajectory: calc.arrowTrajectoryPoints
      }
    })

  return calculatedEntityTarget
}

const currentProjectileDetected: Record<string, Projectil> = {}
export const detectProjectiles = (projectile: string = 'arrow') => {
  const projectiles = Object.values(bot.entities)
    // @ts-ignore PR: https://github.com/PrismarineJS/prismarine-entity/pull/55
    .filter((e) => e.name === projectile && e.type === "projectile")

  const updatedAt = Date.now()

  projectiles.forEach((e) => {
    if (!e.uuid) return
    if (!currentProjectileDetected[e.uuid]) {
      currentProjectileDetected[e.uuid] = {
        uuid: e.uuid,
        enabled: true,
        currentSpeed: 0,
        currentSpeedTime: Date.now(),
        previusPositions: [],
        updatedAt
      }
    } else {
      currentProjectileDetected[e.uuid].updatedAt = updatedAt
    }

    // if (currentProjectileDetected[e.uuid].previusPositions.length > 3) { currentProjectileDetected[e.uuid].previusPositions.shift() }
    currentProjectileDetected[e.uuid].previusPositions.push({
      at: Date.now(),
      pos: e.position.clone()
    })
  })

  Object.entries(currentProjectileDetected)
    .forEach(e => {
      const [uuid, arrow] = e
      if (arrow.updatedAt !== updatedAt) {
        delete currentProjectileDetected[uuid]
      }
    })

  const arrowsInAir: Array<Projectil> = []

  Object.entries(currentProjectileDetected)
    .filter(e => e[1].enabled)
    .forEach((e) => {

      const [uuid, projectil] = e
      const speed = new Vec3(0, 0, 0)

      const previusPositions = projectil.previusPositions

      const totalItemsToCatch = 3
      const start = previusPositions.length >= totalItemsToCatch ? previusPositions.length - totalItemsToCatch : 0
      const previusPositionsTocheck = previusPositions.slice(start)

      for (let i = 1; i < previusPositionsTocheck.length; i++) {
        const pos = previusPositionsTocheck[i]
        const prevPos = previusPositionsTocheck[i - 1]
        speed.x += Math.abs(pos.pos.x - prevPos.pos.x)
        speed.y += Math.abs(pos.pos.y - prevPos.pos.y)
        speed.z += Math.abs(pos.pos.z - prevPos.pos.z)
      }

      speed.x = speed.x / previusPositionsTocheck.length
      speed.y = speed.y / previusPositionsTocheck.length
      speed.z = speed.z / previusPositionsTocheck.length

      const currentSpeed = speed.x + speed.y + speed.z
      if (currentSpeed !== projectil.currentSpeed) {
        projectil.currentSpeed = currentSpeed <= 3 ? currentSpeed : 3
        projectil.currentSpeedTime = Date.now()
      }

      if (projectil.currentSpeed === 0 && Date.now() - projectil.currentSpeedTime > 1500) {
        projectil.enabled = false
      } else {
        arrowsInAir.push(projectil)
      }
    })

  return arrowsInAir

}